[Frontend] Emit incremental compilation info from some emit-module jobs

xymus opened the pull request:
Emit the incremental fine-grained information from the emit-module job to improve the emit-module-separately build mode compatibility with incremental compilation. Note that emit-module doesn't usually parses or type-checks non-inlinable function bodies so it may be unaware of some dependencies.

xymus commented:
@swift-ci Please smoke test

xymus commented:
I'm mostly putting this PR up in support for testing https://github.com/apple/swift-driver/pull/730.

@CodaFi Are there tests in the compiler repo that could be extended to test this or are they on the driver side?

CodaFi commented:
There are some tests under `test/Incremental` that could be extended for frontend-specific testing. We try to keep the integration tests in the driver these days though, yeah.

xymus commented:
apple/swift-driver#730
@swift-ci Please test

davidungar reviewed (approved):
Beautiful!

xymus commented:
@swift-ci Please smoke test

xymus committed:
[Frontend] Emit incr compilation info from emit-module jobs

Enable emitting the module-level incremental fine-grained compilation
information from the emit-module job for incremental compilation to
work with emit-module-separately.

xymus committed:
[AST] Accept SynthesizedFileUnit in populateMemberCache

The differentiation and actor logics insert SynthesizedFileUnit in
SourceFile modules. Accepting these file units in populateMemberCache
allow to cache all the top-level decls of source file modules.

xymus commented:
@swift-ci Please smoke test

CodaFi commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1557,17 +1557,29 @@ static bool performCompileStepsPostSILGen(CompilerInstance &Instance,
-      auto *Mod = MSF.get<ModuleDecl *>();
+
+      StringRef fileFingerprint;
+      if (auto Mod = MSF.dyn_cast<ModuleDecl *>()) {
```
I'm a bit confused as to how we see SourceFiles here. There shouldn't be any primaries, so this path should only be fed by ModuleDecls, right?

xymus replied:
Emit module only has source file inputs that end up here. Do you think that shouldn't happen?

I wouldn't expect this job to be passed the primary flag but I could double check in the tests where this failed.

CodaFi replied:
My expectation is that separate module mode behaves like WMO, not like single-file mode. There shouldn't be primaries here.

xymus replied:
Good point! IIUC a jobs with the action `-emit-module` and using a primary-file would be used to create a partial swiftmodule file, so I'll modify this code to avoid jobs with a SourceFile and it should just generate the graph for module-wide emit-module jobs.

xymus commented on lib/Frontend/Frontend.cpp:
```diff
@@ -343,7 +343,8 @@ void CompilerInstance::setupDependencyTrackerIfNeeded() {
     if (opts.InputsAndOutputs.hasDependencyTrackerPath() ||
-        !opts.IndexStorePath.empty()) {
+        !opts.IndexStorePath.empty() ||
+        opts.RequestedAction == FrontendOptions::ActionType::EmitModuleOnly) {
```
@CodaFi What do you think of forcing dependency tracking for all emit module only jobs? The data structure configured here is used later to generate the fine grained dependency graph which is also now built by default with all emit module only jobs.

CodaFi replied:
This part of the code is about forcing system (read: SDK) dependencies on and off when we import stuff via the clang importer. I don't think you need to amend this logic. The user should be the one in control here with -track-system-dependencies.

xymus replied:
This change is no more necessary when fixing the SourceFile issue above.

xymus commented on lib/AST/Module.cpp:
```diff
@@ -276,8 +276,11 @@ void SourceLookupCache::populateMemberCache(const ModuleDecl &Mod) {
+           isa<SynthesizedFileUnit>(file));
+    SmallVector<Decl *, 8> decls;
+    file->getTopLevelDecls(decls);
+    addToMemberCache(decls);
```
@slavapestov Do you think it's relevant for `SourceLookupCache::populateMemberCache` to cache the decls from `SynthesizedFileUnit` that are part of a module composed otherwise only of `SourceFiles`? The differentiation and actors logic appears to insert `SynthesizedFileUnits` in such modules.

slavapestov replied:
If unqualified lookup can see those decls, I believe they must be present in the cache (since there is no other mechanism to find them), unless we bypass the cache entirely and check each FileUnit in that case (but that would be bad for performance).

xymus replied:
Alright, thanks!

xymus commented:
@swift-ci Please smoke test

xymus commented:
apple/swift-driver#730
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test

xymus commented:
@swift-ci Please smoke test Linux

xymus commented:
@swift-ci Please smoke test Windows
